File: client-admin/types/refund.ts
import { OrderItemSnapshot } from "./order";

export type RefundStatus = "pending" | "processing" | "completed" | "rejected" | "cancelled";

export interface CalculateRefund {
    subtotalRefund: number;
    itemRefund: number;
    shippingRefund: number;
    totalRefund: number;
}

export type ReasonCode =
    | "DEFECTIVE"
    | "WRONG_ITEM"
    | "MISSING_ITEM"
    | "NOT_AS_DESCRIBED"
    | "OTHER";

export interface RefundItemPayload {
    productId: string;
    quantity: number;
}

export interface RefundImagePayload {
    url: string;
    imagePublicId: string;
}

export interface RefundCustomer {
    fullName: string;
    phone: string;
    email: string;
}

export interface BankInfoPayload {
    bankName: string;
    accountNumber: string;
    accountHolder: string;
}

export interface CreateRefundPayload {
    orderCode: string;
    paymentMethod: "cod" | "banking" | "vnpay" | "momo";
    reasonCode: ReasonCode;
    reason?: string;

    note?: string;

    images?: RefundImagePayload[];

    refundDestination?: "original" | "bank";

    refundBankInfo?: BankInfoPayload | null;

    items?: RefundItemPayload[];
}

export interface RefundAmountData {
    subtotalRefund: number;
    itemRefund: number;
    shippingRefund: number;
    totalRefund: number;
}

export interface RefundUserDetail {
    _id: string;
    orderCode: string;
    refundCode: string;

    userId?: string | null;

    type: "full" | "partial";

    reasonCode: ReasonCode;
    reason?: string;

    note?: string;

    status: RefundStatus;

    items: OrderItemSnapshot[];

    images: RefundImagePayload[];

    paymentMethod: "cod" | "banking";

    refundDestination?: "original" | "bank";

    refundBankInfo?: BankInfoPayload | null;

    refundAmountData: RefundAmountData;

    createdAt: string;
    updatedAt: string;
}

export interface RefundSummary {
    _id: string
    orderCode: string
    refundCode: ReasonCode
    reason?: string

    status: RefundStatus
    createdAt: string

    totalRefund: number

    itemCount: number
    type: "full" | "partial"

    reasonCode: string

    refundDestination: "original" | "bank"
}

export type OrderRefundListItem = {
    refundCode: string
    status: RefundStatus
    type: "full" | "partial"
    createdAt: string
    itemCount: number
    totalRefund: number
}

export interface ListRefundItem {
    _id: string;
    orderCode: string;
    refundCode: string;

    status: RefundStatus;
    createdAt: string;
    totalRefund: number;

    items: OrderItemSnapshot[];
    type: "full" | "partial";
    reason: string;
    viewToken: string;
}

export interface RefundableItem {
   productId: string;
    name: string;
    image: string;
    price: number;
    orderedQuantity: number;
    refundedQuantity: number;
    refundableQuantity: number;
}


export interface RefundAdminListItem {
  _id: string

  refundCode: string
  orderCode: string

  customer?: RefundCustomer

  reason?: string
  reasonCode: ReasonCode

  totalRefund: number

  status: RefundStatus

  createdAt: string
}

export interface RefundAdminDetail {
  _id: string
  refundCode: string

  orderId: string
  orderCode: string

  userId?: string | null

  type: "full" | "partial"

  reasonCode: ReasonCode
  reason?: string

  note?: string

  items: OrderItemSnapshot[]
  images: {
    url: string
    imagePublicId: string
  }[]

  paymentMethod: string

  refundDestination: "bank" | "original"

  refundBankInfo?: BankInfoPayload | null

  refundAmountData: RefundAmountData

  status: RefundStatus

  customer: {
    fullName: string
    phone: string
    email: string
  }

  createdAt: string
  updatedAt: string
}

File: server/app/models/refund.py
from app.utils.format_time import to_utc_iso
from bson import ObjectId

def map_refund_for_user(refund: dict) -> dict:
    return {
        "_id": str(refund.get("_id")),
        "orderCode": refund.get("orderCode"),
        "refundCode": refund.get("refundCode"),

        "type": refund.get("type"),
        "status": refund.get("status"),

        "reasonCode": refund.get("reasonCode"),
        "reason": refund.get("reason"),
        "custom": refund.get("custom"),
        "note": refund.get("note"),

        "items": refund.get("items", []),

        "images": refund.get("images", []),

        "refundAmountData": refund.get("refundAmountData"),

        "refundDestination": refund.get("refundDestination"),
        "refundBankInfo": refund.get("refundBankInfo"),

        "createdAt": to_utc_iso(refund.get("createdAt")),
    }

def transform_refund_admin(doc):
    items = doc.get("items", [])

    return {
        "_id": str(doc["_id"]),
        "refundCode": doc.get("refundCode"),

        "orderId": str(doc.get("orderId")) if doc.get("orderId") else None,
        "orderCode": doc.get("orderCode"),

        "userId": str(doc.get("userId")) if doc.get("userId") else None,

        "type": doc.get("type"),

        "reasonCode": doc.get("reasonCode"),
        "reason": doc.get("reason"),
        "note": doc.get("note"),

        "status": doc.get("status"),

        "paymentMethod": doc.get("paymentMethod"),
        "refundDestination": doc.get("refundDestination"),

        "refundBankInfo": doc.get("refundBankInfo"),

        "refundAmountData": doc.get("refundAmountData", {
            "subtotalRefund": 0,
            "itemRefund": 0,
            "shippingRefund": 0,
            "totalRefund": 0
        }),

        "customer":{
            "fullName": doc.get("shippingAddress", {}).get("fullName"),
            "phone": doc.get("shippingAddress", {}).get("phone"),
            "email": doc.get("shippingAddress", {}).get("email"),
        },
            
        "items": [
            {
                "productId": str(item.get("productId")),
                "name": item.get("name"),
                "image": item.get("image", {
                    "url": "",
                    "imagePublicId": None
                }),
                "price": item.get("price", 0),
                "quantity": item.get("quantity", 0)
            }
            for item in items
        ],

        "images": doc.get("images", []),

        "createdAt": doc.get("createdAt"),
        "updatedAt": doc.get("updatedAt"),
    }